[clang][OpenMP] New OpenMP 6.0 threadset clause

Ritanya-B-Bharadwaj opened the pull request:
Initial parsing/sema/codegen support for threadset clause in task and taskloop directives  [Section 14.8 in in OpenMP 6.0 spec]

Ritanya-B-Bharadwaj committed:
[OpenMP] Parsing Support of ThreadSets in Task

Ritanya-B-Bharadwaj committed:
[clang][OpenMP] New OpenMP 6.0 threadset clause

Ritanya-B-Bharadwaj committed:
[clang] [OpenMP] Codegen support for threadset

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -3691,6 +3691,7 @@ CGOpenMPRuntime::emitTaskInit(CodeGenFunction &CGF, SourceLocation Loc,
     DestructorsFlag = 0x8,
     PriorityFlag = 0x20,
     DetachableFlag = 0x40,
+    PoolFlag = 0x80,
```
Is the runtime updated to handle this flag correctly?

Ritanya-B-Bharadwaj replied:
Not yet, I have not added support for it in runtime.

alexey-bataev replied:
At first, this should be made part of the runtime, only after that it can be supported in the compiler

Ritanya-B-Bharadwaj replied:
This is a minimal and safe addition — it doesn’t cause any crashes or affect functionality. It’ll simply take effect once the runtime adds support, and until then, it remains inactive without causing issues.

alexey-bataev replied:
The runtime change is better to implement in a separate PR and post an RFC (maybe?) before.
Many flags can be ignored, just need to be sure that we do not break anything else here and most of the consumers/users are agree with this change

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -3700,6 +3701,11 @@ CGOpenMPRuntime::emitTaskInit(CodeGenFunction &CGF, SourceLocation Loc,
     if (NeedsCleanup)
       Flags = Flags | DestructorsFlag;
   }
+  if (const auto *Clause = D.getSingleClause<OMPThreadsetClause>()) {
```
Better to pass the flag as part of the Data parameter

Ritanya-B-Bharadwaj replied:
Sorry, but could you clarify why that would be necessary? All the flags above seem to be combined using the | operator, so this appears consistent.

alexey-bataev replied:
To keep all the flags and processing in one single place

Ritanya-B-Bharadwaj replied:
Sorry, I’m not sure I fully understand.

Ritanya-B-Bharadwaj replied:
For this flag, I followed the existing pattern where flags are handled directly, not through the data parameter, to stay consistent. Since the other flags are also managed this way, it didn’t seem necessary to change it just for this case. However, if you think using the data parameter is the right approach, I can take it up separately as a new task — to modify the existing features and add support for passing all flags via the data parameter.

alexey-bataev commented on clang/test/OpenMP/task_ast_print.cpp:
```diff
@@ -101,9 +103,11 @@ T tmain(T argc, T *argv) {
   a = 2;
 #pragma omp task default(none), private(argc, b) firstprivate(argv) shared(d) if (argc > 0) final(S<T>::TS > 0) priority(argc) affinity(argc, argv[b:argc], arr[:], ([argc][sizeof(T)])argv)
   foo();
+#ifndef OMP60
```
Why this test should be disabled in 60?

alexey-bataev replied:
Is this expected?

Ritanya-B-Bharadwaj replied:
Yes, according to 6.0 spec - Section 7.6.5 Properties Common to Reduction and induction Clauses
_If the reduction identifier or induction identifier is an implicitly declared reduction identifier or induction identifier or otherwise not an id-expression then it is implicitly converted to one by prepending the keyword operator (for example, + becomes operator+). This conversion is valid for the +, *, /, && and || operators._

alexey-bataev replied:
Hmm, this is strange, that this starts failing. Better to change it to pass under 6.0

Ritanya-B-Bharadwaj replied:
I feel those changes aren't directly related to this PR's functionality. I can take it up separately as an activity to update the lit tests and add 6.0 support. What do you think?

alexey-bataev replied:
Defenitely not a part of this patch

Ritanya-B-Bharadwaj commented:
@alexey-bataev I've addressed your comments. For anything that's out of scope for this patch, I'll handle it in upcoming patches. Let me know if this can be merged.

alexey-bataev reviewed (commented):
Update OpenMPSupport.rst and release notes documents

alexey-bataev commented on clang/test/OpenMP/task_ast_print.cpp:
```diff
@@ -101,9 +103,11 @@ T tmain(T argc, T *argv) {
   a = 2;
 #pragma omp task default(none), private(argc, b) firstprivate(argv) shared(d) if (argc > 0) final(S<T>::TS > 0) priority(argc) affinity(argc, argv[b:argc], arr[:], ([argc][sizeof(T)])argv)
   foo();
+#ifndef OMP60
```
This should be removed from the patch, the test should not be affected

alexey-bataev replied:
Please, modify the test instead by replacing the reduction operaton

dreachem replied:
So you'd like for this part of the test to just use the `+` operator rather than the `-` operator in the clause, thereby making it valid for 6.0 and not requiring the `#ifndef OMP60`? That sounds fine to me, as it doesn't seem that there was any particular significance to using `-` here for what is being tested.

alexey-bataev replied:
Yes, + should be fine

Ritanya-B-Bharadwaj committed:
Adding basic runtime support

alexey-bataev commented:
1. Need to rebase.
2. I'd suggest to split the patch and land the runtime part separately

Ritanya-B-Bharadwaj committed:
Removing runtime changes

Ritanya-B-Bharadwaj commented on clang/docs/OpenMPSupport.rst:
```diff
@@ -364,7 +364,7 @@ implementation.
 | free-agent threads                                          | :none:`unclaimed`         | :none:`unclaimed`         |                                                                          |
 +-------------------------------------------------------------+---------------------------+---------------------------+--------------------------------------------------------------------------+
-| threadset clause                                            | :`worked on`              | :none:`unclaimed`         |                                                                          |
+| threadset clause                                            | :good:`done`              | :none:`unclaimed`         | https://github.com/llvm/llvm-project/pull/135807                         |
```
Noted! Does this seem better?

jprotze replied:
I guess its a POV question :) From an implementer's perspective you are right: the implementation is compliant. From an user's perspective the implementation is at the lowest level of QoI.

I'm not completely sure about the target audience of the page, but I assumed it's part of the user documentation: https://clang.llvm.org/docs/OpenMPSupport.html#openmp-6-0-implementation-details

As a reference: `nowait clause on taskwait` claims _partial_. I think the implementation status is the same: compiler works, and the runtime implementation currently defaults to fallback code that at least provides a compliant solution.

Ritanya-B-Bharadwaj replied:
Agreed! Modified accordingly.

jprotze commented:
I put a comment regarding the runtime API. But, I'm not the person to review the overall frontend patch.

@alexey-bataev since the runtime part has landed, can you have a look on this PR?

Ritanya-B-Bharadwaj committed:
Update clang/lib/CodeGen/CGOpenMPRuntime.cpp

Co-authored-by: Joachim <protze@rz.rwth-aachen.de>

Ritanya-B-Bharadwaj committed:
Update clang/include/clang/AST/OpenMPClause.h

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -3700,6 +3700,7 @@ CGOpenMPRuntime::emitTaskInit(CodeGenFunction &CGF, SourceLocation Loc,
     DestructorsFlag = 0x8,
     PriorityFlag = 0x20,
     DetachableFlag = 0x40,
+    FreeAgentFlag = 0x80,
```
Doe this flag implemented in runtime already?

Ritanya-B-Bharadwaj replied:
Yes @alexey-bataev  - https://github.com/llvm/llvm-project/pull/144409

Ritanya-B-Bharadwaj committed:
Resolving flang build failure

Ritanya-B-Bharadwaj commented:
I’ve made some basic changes to the flang frontend just to fix the build failure. The Clang side is already reviewed and approved. These flang changes are just temporary placeholders until someone takes up proper support for this feature in Fortran.

Ritanya-B-Bharadwaj commented:
ping

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Fixing formatting issues

Ritanya-B-Bharadwaj committed:
Fixing formatting issues

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Update OpenMPKinds.cpp

Ritanya-B-Bharadwaj committed:
Update OMP.td

Ritanya-B-Bharadwaj committed:
Restore original version of OpenMPSupport.rst

jprotze commented:
In general, I think this PR is ready to land, as soon as we get any feedback from flang folks. The clang codegen was reviewed. The runtime implementation is merged. I reviewed the interfacing with the runtime.

@mjklemm the flang implementation in the PR is more a stub than an actual implementation. Can you nevertheless review and comment?

zahiraam commented:
@Ritanya-B-Bharadwaj Can this be pushed along so that it can be merged in? @kparzysz Would you mind taking a look the `Flang` changes please?

mjklemm reviewed (commented):
Flang side changes LGTM. but PR needs some rebasing.

zahiraam commented:
Thanks @mjklemm

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Fixing merge conflicts on OpenMPSupport.rst

kparzysz commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4714,6 +4714,14 @@ struct OmpTaskReductionClause {
+// threadset-clause ->
+//     THREADSET(omp_pool|omp_team)
+struct OmpThreadsetClause {
+  ENUM_CLASS(ThreadsetPolicy, omp_pool, omp_team)
```
Please capitalize: Omp_Pool, Omp_Team.

kparzysz commented on flang/lib/Lower/OpenMP/Clauses.cpp:
```diff
@@ -1426,6 +1426,21 @@ ThreadLimit make(const parser::OmpClause::ThreadLimit &inp,
+  CLAUSET_ENUM_CONVERT( //
+      convert, wrapped::ThreadsetPolicy, Threadset::ThreadsetPolicy,
+      // clang-format off
+      MS(omp_pool, omp_pool)
+      MS(omp_team, omp_team)
```
Capitalize: Omp_Pool, Omp_Team.

kparzysz commented on llvm/include/llvm/Frontend/OpenMP/ClauseT.h:
```diff
@@ -1146,6 +1146,14 @@ struct ThreadsT {
+// V6.0: [14.8] `threadset` clause
+template <typename T, typename I, typename E> //
+struct ThreadsetT {
+  ENUM(ThreadsetPolicy, omp_pool, omp_team);
```
Finally here: Omp_Pool, Omp_Team.

Ritanya-B-Bharadwaj replied:
Done!

Ritanya-B-Bharadwaj committed:
flang changes

kparzysz reviewed (approved):
Flang changes are ok.

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

mjklemm reviewed (approved):
LGTM

Ritanya-B-Bharadwaj committed:
Updating OpenMPSupport.rst format

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Update OpenMPSupport.rst

Ritanya-B-Bharadwaj committed:
Merge branch 'main' into ThreadSet

Ritanya-B-Bharadwaj committed:
Update OpenMPSupport.rst

Ritanya-B-Bharadwaj committed:
Update ClauseT.h

Ritanya-B-Bharadwaj committed:
Update ClauseT.h

h-vetinari commented on clang/docs/OpenMPSupport.rst:
```diff
@@ -1,512 +1,512 @@
-+------------------------------+-----------------------------------------------------------------------------------+--------------------------+--------------------------------------------------------+
-
-.. _Discourse forums (Runtimes - OpenMP category): https://discourse.llvm.org/c/runtimes/openmp/35
+.. raw:: html
```
This did not happen, and now the entire file has been rewritten by https://github.com/llvm/llvm-project/commit/25ece5ba925347a5688f180af3131659948c3828 just due to line-ending changes. You can add `?w=1` to the URL and it'll show the diff without whitespace changes.

In any case, this makes it very hard to use git blame in `clang/docs/OpenMPSupport.rst`, which is something I follow closely.
